src/hardware/vga_draw.cpp

```diff
@@ -1325,24 +1385,54 @@ void VGA_SetupDrawing(uint32_t /*val*/)
 
 		// Adjust the VGA clock frequency based on the Clocking Mode Register's
 		// 9/8 Dot Mode. See Timing Model: https://wiki.osdev.org/VGA_Hardware
-		clock /= (vga.seq.clocking_mode & 1) ? 8 : 9;
+		clock /= vga.seq.clocking_mode.is_eight_dot_mode
+		               ? PixelsPerChar::Eight
+		               : PixelsPerChar::Nine;
 
 		// Adjust the horizontal frequency if in pixel-doubling mode (clock/2)
-		if (vga.seq.clocking_mode & 0x8) {
+		if (vga.seq.clocking_mode.is_pixel_doubling) {
 			htotal *= 2;
 		}
 		vga.draw.address_line_total = (vga.crtc.maximum_scan_line & 0x1f) + 1;
-		if (IS_VGA_ARCH && (svgaCard==SVGA_None) && (vga.mode==M_EGA || vga.mode==M_VGA)) {
-			// vgaonly; can't use with CGA because these use address_line for their
-			// own purposes.
-			// Set the low resolution modes to have as many lines as are scanned - 
-			// Quite a few demos change the max_scanline register at display time
-			// to get SFX: Majic12 show, Magic circle, Copper, GBU, Party91
-			if ( vga.crtc.maximum_scan_line&0x80) vga.draw.address_line_total*=2;
-			vga.draw.double_scan=false;
+		if (IS_VGA_ARCH) {
+			// CRTC Maximum Scan Line Register - Scan Doubling
+			//
+			// When bit 7 is set to 1, 200-scan-line video data is
+			// converted to 400-scan-line output. To do this, the
+			// clock in the row scan counter is divided by 2, which
+			// allows the 200-line modes to be displayed as 400
+			// lines on the display (this is called double scanning;
+			// each line is displayed twice). When this bit is set
+			// to 0, the clock to the row scan counter is equal to
+			// the horizontal scan rate.
+			//
+			// Ref: http://www.osdever.net/FreeVGA/vga/crtcreg.htm#09
+
+			const auto is_scan_doubled = bit::is(vga.crtc.maximum_scan_line,
+			                                     bit::literals::b7);
+
+			if (vga.mode == M_EGA || vga.mode == M_VGA) {
+				// Set the low resolution modes to have as many
+				// lines as are scanned - Quite a few demos
+				// change the max_scanline register at display
+				// time to get SFX: Majic12 show, Magic circle,
+				// Copper, GBU, Party91.
+				//
+				// Note: can't use with CGA because these use
+				// address_line for their own purposes.
+				if (is_scan_doubled) {
+					vga.draw.address_line_total *= 2;
+				}
+				// No-need to artificially perform scan-doubling
+				vga.draw.double_scan = false;
+			}
+			// VGA machine not in EGA or VGA modes
+			else {
+				vga.draw.double_scan = is_scan_doubled;
+			}
+		} else { // non-VGA machine
+			vga.draw.double_scan = (vtotal == 262);
```

Comment:
Explain/extract magic number?

Reply:
Assuming the CGA output is 262 total display lines per frame, which is true if you're actually CGA. Note that one NTSC field is 262.5 scanlines (the half scanline makes it interlaced) but CGA doesn't do the half-a-scanline so technically the composite output is 240p rather than 480i. Although poking at real CGA, the registers function as documented if you want the CRTC to emit the half a scanline to emit 480i even if it just makes the CGA video frame bob up and down slightly.

But then of course that will break for any DOS game that programs custom CGA modes and possibly gets some registers off by 1.

Reply:
The reason CGA composite output works so well with TV sets is because the CGA refresh rate and timing is very close if not exactly the same as NTSC composite video.

Reply:
Thanks for the explanation @joncampbell123; added this as a constant.